utils/fixed_decimal/src/decimal.rs

```diff
@@ -1438,204 +1332,383 @@ impl FixedDecimal {
-    /// dec.expand(0);
-    /// assert_eq!("1", dec.to_string());
-    /// let mut dec = FixedDecimal::from_str("0.5").unwrap();
-    /// dec.expand(0);
-    /// assert_eq!("1", dec.to_string());
-    /// let mut dec = FixedDecimal::from_str("0.6").unwrap();
-    /// dec.expand(0);
-    /// assert_eq!("1", dec.to_string());
-    /// let mut dec = FixedDecimal::from_str("1.5").unwrap();
-    /// dec.expand(0);
-    /// assert_eq!("2", dec.to_string());
+    /// let mut dec = FixedDecimal::from_str("-3.5").unwrap();
+    /// dec.round_with_mode_and_increment(
+    ///     0,
+    ///     RoundingMode::Floor,
+    ///     RoundingIncrement::MultiplesOf1,
+    /// );
+    /// assert_eq!("-4", dec.to_string());
+    /// let mut dec = FixedDecimal::from_str("-3.59").unwrap();
+    /// dec.round_with_mode_and_increment(
+    ///     -1,
+    ///     RoundingMode::Ceil,
+    ///     RoundingIncrement::MultiplesOf2,
+    /// );
+    /// assert_eq!("-3.4", dec.to_string());
+    /// let mut dec = FixedDecimal::from_str("5.455").unwrap();
+    /// dec.round_with_mode_and_increment(
+    ///     -2,
+    ///     RoundingMode::HalfExpand,
+    ///     RoundingIncrement::MultiplesOf5,
+    /// );
+    /// assert_eq!("5.45", dec.to_string());
+    /// let mut dec = FixedDecimal::from_str("-7.235").unwrap();
+    /// dec.round_with_mode_and_increment(
+    ///     -2,
+    ///     RoundingMode::HalfTrunc,
+    ///     RoundingIncrement::MultiplesOf25,
+    /// );
+    /// assert_eq!("-7.25", dec.to_string());
+    /// let mut dec = FixedDecimal::from_str("9.75").unwrap();
+    /// dec.round_with_mode_and_increment(
+    ///     -1,
+    ///     RoundingMode::HalfEven,
+    ///     RoundingIncrement::MultiplesOf5,
+    /// );
+    /// assert_eq!("10.0", dec.to_string());
     /// ```
-    #[inline(never)]
-    pub fn expand(&mut self, position: i16) {
-        self.expand_to_increment_internal(position, NoIncrement)
+    pub fn round_with_mode_and_increment(
+        &mut self,
+        position: i16,
+        mode: RoundingMode,
+        increment: RoundingIncrement,
+    ) {
+        match mode {
+            RoundingMode::Ceil => self.ceil_to_increment_internal(position, increment),
+            RoundingMode::Expand => self.expand_to_increment_internal(position, increment),
+            RoundingMode::Floor => self.floor_to_increment_internal(position, increment),
+            RoundingMode::Trunc => self.trunc_to_increment_internal(position, increment),
+            RoundingMode::HalfCeil => self.half_ceil_to_increment_internal(position, increment),
+            RoundingMode::HalfExpand => self.half_expand_to_increment_internal(position, increment),
+            RoundingMode::HalfFloor => self.half_floor_to_increment_internal(position, increment),
+            RoundingMode::HalfTrunc => self.half_trunc_to_increment_internal(position, increment),
+            RoundingMode::HalfEven => self.half_even_to_increment_internal(position, increment),
+        }
```

Comment:
Thought: I hope the compiler figures out that it should inline `round_with_mode_and_increment` and all these helper functions except the 3 that contain meaningful code (`expand_to_increment_internal`, `floor_to_increment_internal`, and `half_even_to_increment_internal`).

Reply:
I checked the assembly and it inlines everything correctly. `half_even_to_increment_internal` is also inlined but I think is mostly because is only used in that method.